### ISSUE TITLE
[BUG] Truncation issues when using default_evaluator with explainability

### ISSUE THREAD

**Issue by dsgibbons** [mlflow/mlflow#6554]
### Willingness to contribute

Yes. I can contribute a fix for this bug independently.

### MLflow version

eb273d656fa2778cf6c98031fd33c9ee85bec304

### System information

- macOS
- Python 3.7.13

### Describe the problem

#6503 was recently fixed to allow `mlflow.evaluate` to accept non-string column names. Since the bug fix, if the code example from #6503 is changed to set `log_model_explainability=True`, the code will fail when the feature names are truncated as a result of the call to `truncate_str_from_middle`, as ints don't have a `len` method.

Upon coming up with a fix for this, I actually found that if a feature ends up being given the name: `f"{truncated_name}(f_{i+1})` due to truncation,  `_shap_predict_fn` ends up creating a useless prediction function that fills the DataFrame with NaNs. These NaNs don't prevent the associated test in `tests/test_default_evaluator.py` from passing.

TL;DR:
1. Some minor modifications are required to allow non-string column names during evaluation with explainability.
2. The logic used to form the DataFrame in  `_shap_predict_fn` needs tweaking to account for DataFrames with pre-existing column names. 

I will submit a PR that should demonstrate this issue and the intended fix more clearly.

### Code to reproduce issue

```python
import mlflow

import pandas as pd
from sklearn import datasets
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

# prepare train and test data
iris = datasets.load_iris()
X = iris.data
y = iris.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

xy_test = pd.DataFrame(X_test)
xy_test["label"] = y_test

with mlflow.start_run():
    logistic = LogisticRegression(C=1.0)
    logistic.fit(X_train, y_train)
    mod = mlflow.sklearn.log_model(logistic, artifact_path="mymodel")

    results = mlflow.evaluate(
        model=mod.model_uri,
        model_type="classifier",
        data=xy_test,
        evaluator_config={
            "log_model_explainability": True,
        },
        evaluators="default",
        targets="label",
    )
```

### Stack trace

**Error 1 (non-string column names with explainability):**
From `mlflow.utils.string_utils.py`:

```python
Traceback (most recent call last):
  File "a.py", line 30, in <module>
    targets="label",
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/base.py", line 936, in evaluate
    custom_metrics=custom_metrics,
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/base.py", line 644, in _evaluate
    custom_metrics=custom_metrics,
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/default_evaluator.py", line 1031, in evaluate
    return self._evaluate_classifier()
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/default_evaluator.py", line 973, in _evaluate_classifier
    return self._log_and_return_evaluation_result()
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/default_evaluator.py", line 885, in _log_and_return_evaluation_result
    self._log_model_explainability()
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/default_evaluator.py", line 543, in _log_model_explainability
    truncated_feature_names = [truncate_str_from_middle(f, 20) for f in self.feature_names]
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/default_evaluator.py", line 543, in <listcomp>
    truncated_feature_names = [truncate_str_from_middle(f, 20) for f in self.feature_names]
  File "/home/haru/Desktop/repositories/mlflow/mlflow/utils/string_utils.py", line 19, in truncate_str_from_middle
    if len(s) <= max_length:
TypeError: object of type 'int' has no len()
```

**Error 2 (truncation creating NaN DataFrame during evaluation):**
When calling `predict_fn` in `default_evaluator.py`:

```python
Traceback (most recent call last):
  File "a.py", line 30, in <module>
    targets="label",
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/base.py", line 936, in evaluate
    custom_metrics=custom_metrics,
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/base.py", line 644, in _evaluate
    custom_metrics=custom_metrics,
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/default_evaluator.py", line 1032, in evaluate
    return self._evaluate_classifier()
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/default_evaluator.py", line 974, in _evaluate_classifier
    return self._log_and_return_evaluation_result()
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/default_evaluator.py", line 886, in _log_and_return_evaluation_result
    self._log_model_explainability()
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/default_evaluator.py", line 625, in _log_model_explainability
    shap_values = explainer(sampled_X)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/shap/explainers/_exact.py", line 71, in __call__
    batch_size=batch_size, interactions=interactions, silent=silent
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/shap/explainers/_explainer.py", line 260, in __call__
    batch_size=batch_size, outputs=outputs, silent=silent, **kwargs
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/shap/explainers/_exact.py", line 113, in explain_row
    outputs = fm(extended_delta_indexes, zero_index=0, batch_size=batch_size)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/shap/utils/_masked_model.py", line 57, in __call__
    return self._delta_masking_call(masks, zero_index=zero_index, batch_size=batch_size)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/shap/utils/_masked_model.py", line 199, in _delta_masking_call
    outputs = self.model(*subset_masked_inputs)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/shap/models/_model.py", line 21, in __call__
    return np.array(self.inner_model(*args))
  File "/home/haru/Desktop/repositories/mlflow/mlflow/models/evaluation/default_evaluator.py", line 418, in _shap_predict_fn
    return predict_fn(pd.DataFrame(x, columns=feature_names))
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/sklearn/linear_model/_base.py", line 309, in predict
    scores = self.decision_function(X)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/sklearn/linear_model/_base.py", line 284, in decision_function
    X = check_array(X, accept_sparse='csr')
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/sklearn/utils/validation.py", line 63, in inner_f
    return f(*args, **kwargs)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/sklearn/utils/validation.py", line 721, in check_array
    allow_nan=force_all_finite == 'allow-nan')
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/sklearn/utils/validation.py", line 106, in _assert_all_finite
    msg_dtype if msg_dtype is not None else X.dtype)
ValueError: Input contains NaN, infinity or a value too large for dtype('float64').
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by harupy** [comment]
@dsgibbons Thanks for reporting this issue. Looking forward to your PR!

**Comment by harupy** [comment]
Fixed by #6555
